File: android/bridge/src/main/java/ipfs/gomobile/android/InputStreamToGo.java
package ipfs.gomobile.android;

import java.io.InputStream;
import java.util.Arrays;

final class InputStreamToGo implements core.NativeReader {
    private final InputStream inputStream;

    InputStreamToGo(InputStream inputStream) {
        this.inputStream = inputStream;
    }

    public byte[] nativeRead(long size) throws Exception {
        byte[] b = new byte[(int)size];
        while (true) {
            int n = inputStream.read(b);
            if (n == -1) {
                inputStream.close(); // Auto-close inputStream when EOF is reached
                // The Swift/Go interface converts this to nil.
                return new byte[0];
            }
            if (n > 0) {
                if (n == b.length)
                    return b;
                else
                    return Arrays.copyOf(b, n);
            }

            // Iterate to read more than zero bytes.
        }
    }
}

File: ios/Bridge/GomobileIPFS/Sources/InputStreamToGo.swift
//
//  InputStreamToGo.swift
//  GomobileIPFS
//
//  Created by CI Agent on 20/01/2020.
//

import Foundation
import Core

public class InputStreamToGo: NSObject, CoreNativeReaderProtocol {
    private var inputStream: InputStream

    init(_ inputStream: InputStream) {
        self.inputStream = inputStream
        self.inputStream.open()
    }

    public func nativeRead(_ size: Int) throws -> Data {
        let bytes = UnsafeMutablePointer<UInt8>.allocate(capacity: size)
        while true {
            let read = self.inputStream.read(bytes, maxLength: size)

            if read < 0 {
                throw self.inputStream.streamError!
            }
            if read == 0 && self.inputStream.streamStatus == .atEnd {
                self.inputStream.close()
                // The Swift/Go interface converts this to nil.
                return Data(count: 0)
            }
            if read > 0 {
                return Data(bytes: bytes, count: read)
            }

            // Iterate to read more than zero bytes.
        }
    }
}
